backend/store/address.php

```diff
@@ -0,0 +1,332 @@
+
+        if (strlen($in) < 4) {
+            throw new \ValidationException('Name needs to be greater than 3 charactors');
+        }
+
+        $this->name = htmlspecialchars(ucwords($in), ENT_XML1, 'UTF-8');
+    }
+
+    public function set_line1 ($in) {
+        if (!isset($in)) {
+            throw new \ValidationException('Address line 1 is not set');
+        }
+
+        $this->line1 = htmlspecialchars(ucwords($in), ENT_XML1, 'UTF-8');
+    }
+
+    public function set_line2 ($in) {
+        if (!isset($in)) {
+            throw new \ValidationException('Address line 2 is not set');
+        }
+
+        $this->line2 = htmlspecialchars(ucwords($in), ENT_XML1, 'UTF-8');
+    }
+
+    public function set_city ($in) {
+        \validate_string($in);
+
+        if (!preg_match("/^[a-z0-9 ]{3,}$/i", $in)) {
+            throw new \ValidationException('City is not valid');
+        }
+
+        $this->city = htmlspecialchars(ucwords($in), ENT_XML1, 'UTF-8');
+    }
+
+    public function set_state ($in) {
+        if (!isset($in)) {
+            throw new \ValidationException('State code is not set');
+        }
+
+        if (!isset($this->country)) {
+            throw new \ValidationException('Country code is not yet set');
+        }
+
+        $states = get_states($this->country);
+        $in = strtoupper($in);
+
+        if (count($states) === 0) return;
+
+        if (!isset($states[$in])) {
+            throw new \ValidationException('State code does not exist');
+        }
+
+        $this->state = htmlspecialchars($in, ENT_XML1, 'UTF-8');
+    }
+
+    public function set_country ($in) {
+        if (!isset($in) || !is_string($in)) {
+            throw new \ValidationException('Country code is not valid');
+        }
+
+        $countries = get_countries();
+        $in = strtoupper($in);
+
+        if (!isset($countries[$in])) {
+            throw new \ValidationException('Country code does not exist');
+        }
+
+        $this->country = htmlspecialchars($in, ENT_XML1, 'UTF-8');
+    }
+
+    public function set_postal ($in) {
+        if (!isset($in)) {
+            throw new \ValidationException('Postal code is not set');
+        }
+
+        if (!is_string($in) && !is_int($in)) {
+            throw new \ValidationException($m);
+        }
+
+        if (!preg_match("/^[0-9\- ]{4,}$/i", $in)) {
+            throw new \ValidationException($m);
+        }
+
+        $this->postal = htmlspecialchars($in, ENT_XML1, 'UTF-8');
+    }
+
+    public function set_email ($in) {
+        \validate_string($in);
+
+        if (!filter_var($in, FILTER_VALIDATE_EMAIL)) {
+            throw new \ValidationException('Email is not valid');
+        }
+
+        $this->email = htmlspecialchars($in, ENT_XML1, 'UTF-8');
+    }
+
+    public function set_phone ($in) {
+        if (!isset($in)) {
+            throw new \ValidationException('Phone number is not set');
+        }
+
+        if (!preg_match("/^[\+0-9\-\# ]{7,}$/i", $in)) {
```

Comment:
Can we just sanitize this for the user? It's awful to allow the user to type in (555) 123-4567 and then not actually accept it.